### ISSUE TITLE
Split out *standard-acl2-exports* from std/package.lsp

### ISSUE THREAD

**Issue by ragerdl** [acl2/acl2#340]
I'd like to be able to access the variable `*standard-acl2-exports*` without depending explicitly on the `SETS` and `STR` packages.  As such, I'm planning on moving that definition into its own file -- maybe `books/std/package-acl2.lsp` and then ld'ing `books/std/package-acl2.lsp` from `books/std/package.lsp`.

I'd also like to add b* to the list, since we use it so ubiquitously, and I don't know of anyone who's defined their own b* primitive.

If you think the above ideas are misguided, please let me know.

**Comment by jaredcdavis** [comment]
Can you explain your motivation for this a bit more?

Normally package files should just be including `std/portcullis`, and the definition of `*standard-acl2-exports*` will be loaded from there once per session.  If you split it out into a separate file and, e.g., `ld` it from other files, then it'll get processed more than once...

**Comment by ragerdl** [comment]
Is the vision that any books that follow the Standard Approach include `std/portcullis` as part of their cert.acl2 file?  Maybe I'll add this to best practices if it's not already there.

I agree that, hypothetically speaking, processing a file more than once is a bit annoying.  OTOH, the wall-clock time for such double/triple/etc-processing is probably pretty low.  Does such double-inclusion result in a messy history?

**Comment by jaredcdavis** [comment]
This is kind of complicated.  the TL;DR version is:

 - You **should** use `(include-book "[...]/portcullis")` in your `cert.acl2` files
 - You **should not** use `(ld "[...]/package.lsp")` in your `cert.acl2` files.

I think this is documented in the `WORKING-WITH-PACKAGES` topic, but of course maybe that topic could be elaborated on.

Here is a longer explanation.

If you write a book like this:

```
;;; mybook.lisp
(in-package "ACL2")
(include-book "std/util/defaggregate" :dir :system)
(defun f (x) x)
```

You will find that you can certify it without having an explicit `cert.acl2` file.  However, if you write a book like this:

```
;;; mybook2.lisp
(in-package "ACL2")
(include-book "std/util/defaggregate" :dir :system)
(defun std::f (x) x)
```

Then you won't be able to certify it without a `cert.acl2` file that loads the `std` package.

Why not?  I think this is because ACL2 is going to try to read your whole book first and then process its forms, and to read the book successfully it will have to know about the `std` package before it starts reading the forms.

So, you need the `std` package.  You could get it in many ways.  Probably the most straightforward choices are to add either of the following to your `cert.acl2` file:

 - `(include-book "std/portcullis" :dir :system)`, or
 - `(ld "std/package.lsp" :dir :system)`

But it is better to do the `include-book`.

If you put the `include-book` into your `cert.acl2` file, then all of your own books will have `(include-book "std/portcullis" ...)` in their portcullis.  You might think that, when you include many of your own books, this will cause you to re-process `std/portcullis` over and over again.  But ACL2 is smart enough to see that those includes are redundant, so all of these includes (except for the first) are just become fast redundant events that don't have to re-process anything that's inside of `std/portcullis`.

In contrast, if you put `(ld "std/package.lsp" :dir :system)` in your `cert.acl2` file, then things do not work out as well.  In this case, each of your books will have the `ld` form in their portcullis.  When you include several of your books together, ACL2 will have to process all of these `ld` forms.  Each of these forms will cause the `std/package.lsp` file to be loaded and processed, over and over.

**Comment by solswords** [comment]
Jared Davis <notifications@github.com> wrote:

>
>  You will find that you can certify it without having an explicit
> cert.acl2 file. However, if you write a book like this:
>
> ;;; mybook2.lisp
> (in-package "ACL2")
> (include-book "std/util/defaggregate" :dir :system)
> (defun std::f (x) x)
>
> Then you won't be able to certify it without a cert.acl2 file that loads
> the std package.
>
Nowadays the build system should actually be able to take care of this case
without a cert.acl2 file -- as long as the package is defined in the
portcullis of some sub-book, it shouldn't be necessary.  (The build system
LDs the port files of all included books before starting certification --
this means you shouldn't have to worry about using packages, except in the
case that your book uses a package that none of its sub-books are using.)

The point about preferring (include-book "portcullis") to (ld
"package.lsp") still stands -- it's much better to check redundancy of an
include-book than to check redundancy of several defpkg forms, which is
what happens if you use LD instead.  (The build system's LDs tend to work
out well as long as the things in the cert.acl2 files at the "leaves" are
wrapped up in high-level include-books.)

**Comment by ragerdl** [comment]
At this point, I've figured out that this is a slightly grey issue, and I'm just going to leave things as the status quo.  I think splitting out the constant `*standard-acl2-exports*` could be a nice design improvement, but there isn't really any practical benefit to such a change.  Furthermore, there is a non-trivial argument against splitting out the constant, which is that the exclusion of symbols like `union` only make sense to exclude in the sense of `std`.

Just for clarity, I will restate the original proposed plan (which is not going to be enacted): To put `(ld "std/acl2-package.lsp")` inside `books/std/package.lsp`.  This means I could `ld` `acl2-package.lsp` from other `package.lsp` files, as would be consistent with the Books Best Practices.
